Tag API. (#679)
Unleash/unleash: docs/api/admin/tags-api.md

---
id: tags
title: /api/admin/tags
---

> > In order to access the admin api endpoints you need to identify yourself. If you are using the `unsecure` authentication method, you may use [basic authentication](https://en.wikipedia.org/wiki/Basic_access_authentication) to ientify yourself.

Reviewer: ```suggestion
> In order to access the admin API endpoints you need to identify yourself. If you are using the `insecure` authentication method, you may use [basic authentication](https://en.wikipedia.org/wiki/Basic_access_authentication) to identify yourself.
```

# Create a new tag

> `POST http://unleash.host.com/api/admin/tags`

Reviewer: ```suggestion
`POST https://unleash.host.com/api/admin/tags`
```

Used to create a new tag

Used by the admin-dashboard to create a new tag

**Body:**

```json
{
  "type": "simple",
  "value": "Team-Green"
}
```

## Success

    - Returns _200-OK_ if the tag was created successfully

## Failures

>     - Returns 404 and the following error if the _type_ was not found

Reviewer: ```suggestion
    - Returns 404 and the following error if the `type` was not found
```

```json
{ "error": "Tag type not found" }
```

### Notes

- `type` must exist in tagtypes

# List tags

> `GET http://unleash.host.com/api/admin/tags`

Reviewer: ```suggestion
`GET https://unleash.host.com/api/admin/tags`
```

This endpoint is the one all admin UIs should use to fetch all available tags from the _unleash_server_. The response returns all tags.

**Example response:**

```json
{
  "version": 1,
  "tags": [
    {
      "id": 5,
      "value": "Team-Red",
      "type": "simple"
    },
    {
      "id": 9,
      "value": "Team-Green",
      "type": "simple"
    },
    {
      "id": 19,
      "value": "DecemberExperiment",
      "type": "simple"
    }
  ]
}
```

# Update a tag

> `PUT: http://unleash.host.com/api/admin/tags/:tagId`

Reviewer: ```suggestion
`PUT: https://unleash.host.com/api/admin/tags/:tagId`
```

**Body**

```json
{
  "value": "New value for tag"
}
```

# Add tags to a feature

> `POST: http://unleash.host.com/api/admin/tag/:featureName`

Reviewer: ```suggestion
`POST: https://unleash.host.com/api/admin/tag/:featureName`
```

**_Body _**

```json
{
  "tags": [5, 9]
}
```

Connects `featureName` to list of tags by id

# Fetching Tag types

> `GET: http://unleash.host.com/api/admin/tag-types`

Reviewer: ```suggestion
`GET: https://unleash.host.com/api/admin/tag-types`
```

> Used to fetch all types the server knows about. This endpoint is the one all admin ui should use to fetch all available tag types from the _unleash-server_. The response returns all tag types. Any server wil have _at least_ one configured tag type (the `simple` type). A tag type will have a `type`, `description`, `icon` map

Reviewer: ```suggestion
Used to fetch all types the server knows about. This endpoint is the one all admin UI should use to fetch all available tag types from the _unleash-server_. The response returns all tag types. Any server will have _at least_ one configured tag type (the `simple` type). A tag type will be a map of `type`, `description`, `icon`
```

**Example response:**

```json
{
  "version": 1,
  "types": [
    {
      "name": "simple",
      "description": "Arbitrary tags. Used to simplify filtering of features",
      "icon": "#"
    }
  ]
}
```

# Create a new tag type

> `POST: http://unleash.host.com/api/admin/tagtypes`

Reviewer 2: rename to `tag-types`?

Author: yes. Thanks. missed that

> `POST: http://unleash.host.com/api/admin/tag-types`

Reviewer: ```suggestion
`POST: https://unleash.host.com/api/admin/tag-types`
```

> Used to register a new tagtype. This endpoint should be used to inform the server about a new type of tags.

Reviewer: ```suggestion
Used to register a new tag type. This endpoint should be used to inform the server about a new type of tags.
```

**Body:**

```json
{
  "name": "tagtype",
  "description": "Purpose of tag type",
  "icon": "Either an URL to icon or a simple prefix string for tag"
}
```

**Notes:**

> - _name_ **must be unique**, otherwise you will get a _403-response_.

Reviewer: ```suggestion
- `name` **must be unique**, otherwise you will get a _403-response_.
```

Returns 200-OK if the tag type was created successfully

### Update tag type

> `PUT: http://unleash.host.com/api/admin/tag-types/:typeName`

Reviewer: ```suggestion
`PUT: https://unleash.host.com/api/admin/tag-types/:typeName`
```

**Body:**

```json
{
  "description": "New description",
  "icon": "New icon"
}
```

### Deleting a tag type

> `DELETE: http://unleash.host.com/api/admin/tag-types/:typeName`

Reviewer: ```suggestion
`DELETE: https://unleash.host.com/api/admin/tag-types/:typeName`
```

---
id: tags
title: /api/admin/tags
---

> In order to access the admin API endpoints you need to identify yourself. If you are using the `insecure` authentication method, you may use [basic authentication](https://en.wikipedia.org/wiki/Basic_access_authentication) to identify yourself.

### Create a new tag

`POST https://unleash.host.com/api/admin/tags`

> Creates a new tag without connecting it to any other object, can be helpful to build an autocomplete list.

Reviewer 2: I wonder if tags also should expose a "autocomplete" endpoint. E.g 
`POST https://unleash.host.com/api/admin/tags/autocomplete?type=simple&value=a` 

Should return all tag values for a tag-type=simple and value starts with "a" (ignoring case).

Author: As a POST or as a GET?

      "type": "simple"
    },
    {
      "id": 9,
      "value": "Team-Green",
      "type": "simple"
    },
    {
      "id": 19,
      "value": "DecemberExperiment",
      "type": "simple"
    }
  ]
}
```

### Fetching Tag types

`GET: http://unleash.host.com/api/admin/tag-types`

Used to fetch all types the server knows about. This endpoint is the one all admin ui should use to fetch all available tag types from the _unleash-server_. The response returns all tag types. Any server wil have _at least_ one configured tag type (the `simple` type). A tag type will have a `type`, `description`, `icon` map

**Example response:**

```json
{
  "version": 1,
  "tagTypes": [
    {
      "name": "simple",
      "description": "Arbitrary tags. Used to simplify filtering of features",
      "icon": "#"
    }
  ]
}
```

### Get a single tag type

`GET: https://unleash.host.com/api/admin/tag-types/simple`

Used to fetch details about a specific tag-type. This is mostly provded to make it easy to debug the API and should not be used by the client implementations.

**Example response:**

```json
{
  "version": 1,
  "tagType": {
    "name": "simple",
    "description": "Some description",

>     "icon": "Some icon",

Reviewer 2: what do we support in regards of icons? A valid [material-ui icon name](https://material.io/resources/icons/?style=baseline)? A url? An inline svg?

Author: After testing a bit in unleash-frontend with Chip and ChipContact from react-mdl, ChipContact does not display an <img> object, so I've fallen back to exposing either icon or an upcased first letter from name.

 Maybe we should try to look at allowing material-ui icon names? 
Allowing arbitrary URLs does necessitate a better validation of the input to make sure we don't open an attack vector.

}
```

### Get a single tag type

`GET: https://unleash.host.com/api/admin/tag-types/simple`

Used to fetch details about a specific tag-type. This is mostly provded to make it easy to debug the API and should not be used by the client implementations.

**Example response:**

```json
{
  "version": 1,
  "tagType": {
    "name": "simple",
    "description": "Some description",
    "icon": "Some icon"

>   }

Reviewer 2: should we also have a "createdAt" field?

Author: Yes, good point. I'll add it

Author: Do we want it for tags as well?

Reviewer 2: I am unsure of the value, but it does not hurt to at least have the column in the dB, we can expose it later. 

It might be more valuable to have it on the coupling table (`feature_tag`).

Reviewer 3: They're different cases really; when the tag was created, vs when a given feature was tagged.

Reviewer 2: this is a good point. I think we can have both, I think it can be useful in some scenarios.
